libraries/chain/include/eosio/chain/finality/qc.hpp

```diff
@@ -0,0 +1,261 @@
+                                  uint64_t weight);
+
+      // called by add_vote, already protected by mutex
+      vote_status add_weak_vote(size_t index,
+                                const bls_signature& sig,
+                                uint64_t weight);
+
+      bool is_quorum_met_no_lock() const;
+      qc_sig_t to_valid_qc_sig() const;
+   };
+
+   // finalizer authority of strong, weak, or missing votes
+   struct qc_vote_metrics_t {
+      std::set<finalizer_authority_ptr> strong_votes;
+      std::set<finalizer_authority_ptr> weak_votes;
+      std::set<finalizer_authority_ptr> missing_votes;
+   };
+
+   /**
+    * All public methods are thread-safe, pending_policy_sig optional set at construction time.
+    */
+   class open_qc_t {
+   public:
+      open_qc_t(const finalizer_policy_ptr& active_finalizer_policy,
+                const finalizer_policy_ptr& pending_finalizer_policy)
+         : active_finalizer_policy(active_finalizer_policy)
+         , pending_finalizer_policy(pending_finalizer_policy)
+         , active_policy_sig{open_qc_sig_t{active_finalizer_policy->finalizers.size(),
+                                              active_finalizer_policy->threshold,
+                                              active_finalizer_policy->max_weak_sum_before_weak_final()}}
+         , pending_policy_sig{!pending_finalizer_policy
+                                 ? std::optional<open_qc_sig_t>{}
+                                 : std::optional<open_qc_sig_t>{std::in_place,
+                                                                   pending_finalizer_policy->finalizers.size(),
+                                                                   pending_finalizer_policy->threshold,
+                                                                   pending_finalizer_policy->max_weak_sum_before_weak_final()}}
+      {}
+
+      open_qc_t() = default;
+
+      std::optional<qc_t> get_best_qc(block_num_type block_num) const;
+      // verify qc against active and pending policy
+      void verify_qc(const qc_t& qc, const digest_type& strong_digest, const weak_digest_t& weak_digest) const;
+      qc_vote_metrics_t vote_metrics(const qc_t& qc) const;
+      // return qc missing vote's finalizers
+      std::set<finalizer_authority_ptr> missing_votes(const qc_t& qc) const;
+      void set_received_qc(const qc_t& qc);
+      bool received_qc_is_strong() const;
+      vote_status aggregate_vote(uint32_t connection_id, const vote_message& vote,
+                                 block_num_type block_num, std::span<const uint8_t> finalizer_digest);
+      has_vote_status_t has_voted(const bls_public_key& key) const;
+      bool is_quorum_met() const;
+
+   private:
+      friend struct fc::reflector<open_qc_t>;
+      finalizer_policy_ptr         active_finalizer_policy;  // not modified after construction
+      finalizer_policy_ptr         pending_finalizer_policy; // not modified after construction
```

Comment:
why not make them const if they are not modified after construction

Reply:
That was my first thought as well. Unfortunately, it is used in `FC_REFLECT`
```c++
/home/heifner/ext/leap/libraries/libfc/include/fc/io/raw.hpp:250:8: error: static assertion failed due to requirement '!std::is_same_v<const std::shared_ptr<eosio::chain::finalizer_policy>, const std::shared_ptr<eosio::chain::finalizer_policy>>': can't unpack const type
       static_assert(not std::is_same_v<const T, const T>, "can't unpack const type");
```

Reply:
Maybe add a comment saying why `const` is not used.